service/storage_service.cc

```diff
@@ -7487,21 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                                                                    tm->get_my_id(), _snitch.local()->get_location(),
+                                                                    "Tablet migration",
+                                                                    streaming::stream_reason::tablet_migration,
+                                                                    topo_guard,
+                                                                    std::move(tables));
+                tm = nullptr;
+                streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
+                        _gossiper.get_unreachable_members()));
+
+                std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
+                ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
+                streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+                co_await streamer->stream_async();
+                break;
+            }
+            case locator::tablet_transition_kind::rebuild: {
+                // The following streams data from all replicas which are not marked as excluded
+                // Into the pending replica.
+                // FIXME: Use repair
+                auto excluded = _topology_state_machine._topology.get_excluded_nodes();
+                auto& table = _db.local().find_column_family(tablet.table);
+                std::vector<sstring> tables = {table.schema()->cf_name()};
+                auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
+                                                                    guard.get_abort_source(),
+                                                                    tm->get_my_id(), _snitch.local()->get_location(),
+                                                                    "Tablet rebuild",
+                                                                    streaming::stream_reason::tablet_rebuild,
+                                                                    topo_guard,
+                                                                    std::move(tables));
+                tm = nullptr;
+                streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
+                        _gossiper.get_unreachable_members()));
+
+                std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
+                for (auto r : trinfo->next) {
+                    if (r.host != trinfo->pending_replica.host && !excluded.contains(raft::server_id(r.host.uuid()))) {
+                        ranges_per_endpoint[host2ip(r.host)].emplace_back(range);
+                    }
+                }
+                streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+                co_await streamer->stream_async();
```

Comment:
Can't we fold these two together? Doesn't get_migration_streaming_info() abstract all the differences between repair and rebuild?

Reply:
They would have to be unfolded soon when we do repair instead.

Reply:
It could still be common code. We pick one of the set to stream from, then repair if the set size was larger then one. For migration, that doesn't happen. Also true for rebuild with RF=2, so common code here is more general.